File: Example/FWFramework/Classes/Module/Test/Toolkit/TestAnimationController.m
//
//  TestAnimationController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestAnimationController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestAnimationView : UIView

@property (nonatomic, strong) UIView *bottomView;
@property (nonatomic, assign) NSInteger transitionType;

@end

@implementation TestAnimationView

- (instancetype)initWithTransitionType:(NSInteger)transitionType
{
    self = [super initWithFrame:CGRectZero];
    if (self) {
        self.transitionType = transitionType;
        if (self.transitionType > 8) {
            self.backgroundColor = [UIColor clearColor];
        } else {
            self.backgroundColor = [UIColor fw_colorWithHex:0x000000 alpha:0.5];
        }
        
        self.bottomView = [UIView new];
        self.bottomView.backgroundColor = [UIColor whiteColor];
        [self addSubview:self.bottomView];
        if (self.transitionType == 6 || self.transitionType == 9) {
            self.bottomView.fw_layoutChain.left().right().bottom().height(FWScreenHeight / 2);
        } else {
            self.bottomView.fw_layoutChain.center().width(300).height(200);
        }
        
        FWWeakifySelf();
        [self fw_addTapGestureWithBlock:^(id  _Nonnull sender) {
            FWStrongifySelf();
            if (self.transitionType > 8) {
                [self.fw_viewController dismissViewControllerAnimated:YES completion:nil];
                return;
            }
            
            if (self.transitionType == 6) {
                [self fw_setPresentTransition:FWAnimatedTransitionTypeDismiss contentView:self.bottomView completion:nil];
            } else if (self.transitionType == 7) {
                [self fw_setAlertTransition:FWAnimatedTransitionTypeDismiss completion:nil];
            } else {
                [self fw_setFadeTransition:FWAnimatedTransitionTypeDismiss completion:nil];
            }
        }];
    }
    return self;
}

- (void)showInViewController:(UIViewController *)viewController
{
    if (self.transitionType > 8) {
        UIViewController *wrappedController = [self fw_wrappedTransitionController:YES];
        if (self.transitionType == 9) {
            [wrappedController fw_setPresentTransition:nil];
        } else if (self.transitionType == 10) {
            [wrappedController fw_setAlertTransition:nil];
        } else {
            [wrappedController fw_setFadeTransition:nil];
        }
        [viewController presentViewController:wrappedController animated:YES completion:nil];
        return;
    }
    
    [self fw_transitionToController:viewController pinEdges:YES];
    if (self.transitionType == 6) {
        [self fw_setPresentTransition:FWAnimatedTransitionTypePresent contentView:self.bottomView completion:nil];
    } else if (self.transitionType == 7) {
        [self fw_setAlertTransition:FWAnimatedTransitionTypePresent completion:nil];
    } else {
        [self fw_setFadeTransition:FWAnimatedTransitionTypePresent completion:nil];
    }
}

@end

@interface TestAnimationChildController : UIViewController <FWViewController>

@property (nonatomic, strong) UIView *bottomView;
@property (nonatomic, assign) NSInteger transitionType;

@end

@implementation TestAnimationChildController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.fw_navigationBarHidden = YES;
    self.view.backgroundColor = [UIColor clearColor];
    FWWeakifySelf();
    [self.view fw_addTapGestureWithBlock:^(id  _Nonnull sender) {
        FWStrongifySelf();
        [self fw_closeViewControllerAnimated:YES];
    }];
    
    self.bottomView = [UIView new];
    self.bottomView.backgroundColor = [UIColor whiteColor];
    [self.view addSubview:self.bottomView];
    if (self.transitionType == 0 || self.transitionType == 3) {
        self.bottomView.fw_layoutChain.left().right().bottom().height(FWScreenHeight / 2);
    } else {
        self.bottomView.fw_layoutChain.center().width(300).height(200);
    }
    
    UIButton *button = [UIButton new];
    [button setTitleColor:UIColor.blackColor forState:UIControlStateNormal];
    NSString *buttonTitle = self.navigationController ? @"支持push" : @"不支持push";
    [button setTitle:buttonTitle forState:UIControlStateNormal];
    [self.bottomView addSubview:button];
    [button fw_addTouchBlock:^(id  _Nonnull sender) {
        FWStrongifySelf();
        TestAnimationChildController *animationController = [TestAnimationChildController new];
        animationController.transitionType = self.transitionType;
        [self.navigationController pushViewController:animationController animated:YES];
    }];
    button.fw_layoutChain.center();
}

- (void)showInViewController:(UIViewController *)viewController
{
    if (self.transitionType == 0) {
        [self fw_setPresentTransition:nil];
    } else if (self.transitionType == 1) {
        [self fw_setAlertTransition:nil];
    } else {
        [self fw_setFadeTransition:nil];
    }
    [viewController presentViewController:self animated:YES completion:nil];
}

- (void)showInNavigationController:(UIViewController *)viewController
{
    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:self];
    if (self.transitionType == 3) {
        [nav fw_setPresentTransition:nil];
    } else if (self.transitionType == 4) {
        [nav fw_setAlertTransition:nil];
    } else {
        [nav fw_setFadeTransition:nil];
    }
    [viewController presentViewController:nav animated:YES completion:nil];
}

@end

@interface TestAnimationController () <FWViewController>

FWLazyProperty(UIView *, animationView);

@end

@implementation TestAnimationController {
    NSInteger animationIndex_;
}

FWDefLazyProperty(UIView *, animationView, {
    _animationView = [[UIView alloc] initWithFrame:CGRectMake(FWScreenWidth / 2.0 - 75.0, 20, 150, 200)];
    _animationView.backgroundColor = [UIColor redColor];
    [self.view addSubview:_animationView];
});

- (void)didInitialize
{
    self.fw_extendedLayoutEdge = UIRectEdgeBottom;
}

- (void)setupSubviews
{
    UIButton *button = [AppTheme largeButton];
    [button setTitle:@"转场动画" forState:UIControlStateNormal];
    [button fw_addTouchTarget:self action:@selector(onPresent)];
    [self.view addSubview:button];
    [button fw_pinEdgeToSuperview:NSLayoutAttributeBottom withInset:15];
    [button fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    
    UIButton *button2 = [AppTheme largeButton];
    [button2 setTitle:@"切换拖动" forState:UIControlStateNormal];
    [button2 fw_addTouchTarget:self action:@selector(onDrag:)];
    [self.view addSubview:button2];
    [button2 fw_pinEdge:NSLayoutAttributeBottom toEdge:NSLayoutAttributeTop ofView:button withOffset:-15];
    [button2 fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    
    UIButton *button3 = [AppTheme largeButton];
    [button3 setTitle:@"切换动画" forState:UIControlStateNormal];
    [button3 fw_addTouchTarget:self action:@selector(onAnimation:)];
    [self.view addSubview:button3];
    [button3 fw_pinEdge:NSLayoutAttributeBottom toEdge:NSLayoutAttributeTop ofView:button2 withOffset:-15];
    [button3 fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
}

#pragma mark - Action

- (void)onPresent
{
    FWWeakifySelf();
    [self fw_showSheetWithTitle:nil message:nil cancel:@"取消" actions:@[@"VC present", @"VC alert", @"VC fade", @"nav present", @"nav alert", @"nav fade", @"view present", @"view alert", @"view fade", @"wrapped present", @"wrapped alert", @"wrapped fade"] actionBlock:^(NSInteger index) {
        FWStrongifySelf();
        if (index < 3) {
            TestAnimationChildController *animationController = [TestAnimationChildController new];
            animationController.transitionType = index;
            [animationController showInViewController:self];
        } else if (index < 6) {
            TestAnimationChildController *animationController = [TestAnimationChildController new];
            animationController.transitionType = index;
            [animationController showInNavigationController:self];
        } else {
            TestAnimationView *animationView = [[TestAnimationView alloc] initWithTransitionType:index];
            [animationView showInViewController:self];
        }
    }];
}

- (void)onAnimation:(UIButton *)sender
{
    animationIndex_++;
    
    NSString *title = nil;
    if (animationIndex_ == 1) {
        title = @"Push.FromTop";
        [self.animationView fw_addTransitionWithType:kCATransitionPush
                                            subtype:kCATransitionFromTop
                                     timingFunction:kCAMediaTimingFunctionEaseInEaseOut
                                           duration:1.0
                                         completion:NULL];
    }
    
    if (animationIndex_ == 2) {
        title = @"CurlUp";
        [self.animationView fw_addAnimationWithCurve:UIViewAnimationCurveEaseInOut
                                         transition:UIViewAnimationTransitionCurlUp
                                           duration:1.0
                                         completion:NULL];
    }
    
    if (animationIndex_ == 3) {
        title = @"transform.rotation.y";
        [self.animationView fw_addAnimationWithKeyPath:@"transform.rotation.y"
                                            fromValue:@(0)
                                              toValue:@(M_PI)
                                             duration:1.0
                                           completion:NULL];
    }
    
    if (animationIndex_ == 4) {
        title = @"Shake";
        [self.animationView fw_shakeWithTimes:10 delta:0 duration:0.1 completion:NULL];
    }
    
    if (animationIndex_ == 5) {
        title = @"Alpha";
        [self.animationView fw_fadeWithAlpha:0.0 duration:1.0 completion:^(BOOL finished) {
            [self.animationView fw_fadeWithAlpha:1.0 duration:1.0 completion:NULL];
        }];
    }
    
    if (animationIndex_ == 6) {
        title = @"Rotate";
        [self.animationView fw_rotateWithDegree:180 duration:1.0 completion:NULL];
    }
    
    if (animationIndex_ == 7) {
        title = @"Scale";
        [self.animationView fw_scaleWithScaleX:0.5 scaleY:0.5 duration:1.0 completion:^(BOOL finished) {
            [self.animationView fw_scaleWithScaleX:2.0 scaleY:2.0 duration:1.0 completion:NULL];
        }];
    }
    
    if (animationIndex_ == 8) {
        title = @"Move";
        CGPoint point = self.animationView.frame.origin;
        [self.animationView fw_moveWithPoint:CGPointMake(10, 10) duration:1.0 completion:^(BOOL finished) {
            [self.animationView fw_moveWithPoint:point duration:1.0 completion:NULL];
        }];
    }
    
    if (animationIndex_ == 9) {
        title = @"Frame";
        CGRect frame = self.animationView.frame;
        [self.animationView fw_moveWithFrame:CGRectMake(10, 10, 50, 50) duration:1.0 completion:^(BOOL finished) {
            [self.animationView fw_moveWithFrame:frame duration:1.0 completion:NULL];
        }];
    }
    
    if (animationIndex_ == 10) {
        title = @"切换动画";
        animationIndex_ = 0;
    }
    
    if (title) {
        [sender setTitle:title forState:UIControlStateNormal];
    }
}

- (void)onDrag:(UIButton *)sender
{
    if (!self.animationView.fw_dragEnabled) {
        self.animationView.fw_dragEnabled = YES;
        self.animationView.fw_dragLimit = CGRectMake(0, 0, FWScreenWidth, FWScreenHeight - FWNavigationBarHeight - FWStatusBarHeight);
    } else {
        self.animationView.fw_dragEnabled = NO;
    }
}

@end

File: Example/FWFramework/Classes/Module/Test/Toolkit/TestAnimationController.swift
//
//  TestAnimationController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/21.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework
#if DEBUG
import FWDebug
#endif

class TestAnimationController: UIViewController, ViewControllerProtocol {
    
    var animationIndex: Int = 0
    
    lazy var animationView: UIView = {
        let result = UIView()
        result.frame = CGRect(x: APP.screenWidth / 2 - 75, y: 170, width: 150, height: 200)
        result.backgroundColor = UIColor.red
        return result
    }()
    
    func didInitialize() {
        app.extendedLayoutEdge = .bottom
    }
    
    func setupNavbar() {
        app.setRightBarItem(Icon.iconImage("zmdi-var-bug", size: 24)) { _ in
            guard UIWindow.app.main?.app.subview(tag: 1000) == nil else {
                UIWindow.app.main?.app.subview(tag: 1000)?.removeFromSuperview()
                return
            }
            
            let circleView = UIView(frame: CGRect(x: APP.screenWidth / 2 - 25, y: APP.screenHeight / 2 - 25, width: 50, height: 50))
            circleView.tag = 1000
            circleView.app.setCornerRadius(25)
            circleView.backgroundColor = UIColor.app.randomColor
            circleView.app.dragEnabled = true
            circleView.app.dragLimit = CGRect(x: 0, y: 0, width: APP.screenWidth, height: APP.screenHeight)
            circleView.app.isPenetrable = true
            UIWindow.app.main?.addSubview(circleView)
            
            let clickView = UIImageView(frame: CGRect(x: 10, y: 10, width: 30, height: 30))
            clickView.isUserInteractionEnabled = true
            clickView.image = UIImage.app.appIconImage()
            clickView.app.setCornerRadius(15)
            clickView.app.addTapGesture { _ in
                #if DEBUG
                FWDebugManager.sharedInstance().toggle()
                #endif
            }
            clickView.addGestureRecognizer(UILongPressGestureRecognizer.app.gestureRecognizer(block: { _ in
                UIWindow.app.main?.app.subview(tag: 1000)?.removeFromSuperview()
            }))
            circleView.addSubview(clickView)
        }
    }
    
    func setupSubviews() {
        let button = AppTheme.largeButton()
        button.setTitle("转场动画", for: .normal)
        button.app.addTouch(target: self, action: #selector(onPresent))
        view.addSubview(button)
        button.app.layoutChain
            .bottom(15)
            .centerX()
        
        let button2 = AppTheme.largeButton()
        button2.setTitle("切换拖动", for: .normal)
        button2.app.addTouch(target: self, action: #selector(onDrag(_:)))
        view.addSubview(button2)
        button2.app.layoutChain
            .bottom(toViewTop: button, offset: -15)
            .centerX()
        
        let button3 = AppTheme.largeButton()
        button3.setTitle("切换动画", for: .normal)
        button3.app.addTouch(target: self, action: #selector(onAnimation(_:)))
        view.addSubview(button3)
        button3.app.layoutChain
            .bottom(toViewTop: button2, offset: -15)
            .centerX()
        
        view.addSubview(animationView)
    }
    
    @objc func onPresent() {
        app.showSheet(title: nil, message: nil, cancel: "取消", actions: ["VC present", "VC alert", "VC fade", "nav present", "nav alert", "nav fade", "view present", "view alert", "view fade", "wrapped present", "wrapped alert", "wrapped fade"], currentIndex: -1) { [weak self] index in
            guard let self = self else { return }
            if index < 3 {
                let vc = TestAnimationChildController()
                vc.transitionType = index
                vc.show(in: self)
            } else if index < 6 {
                let vc = TestAnimationChildController()
                vc.transitionType = index
                vc.showNav(in: self)
            } else {
                let view = TestAnimationView(transitionType: index)
                view.show(in: self)
            }
        }
    }
    
    @objc func onAnimation(_ sender: UIButton) {
        self.animationIndex += 1
        var title: String? = nil
        if animationIndex == 1 {
            title = "Push.FromTop"
            animationView.app.addTransition(type: .push, subtype: .fromTop, timingFunction: .init(name: .easeInEaseOut), duration: 1.0)
        } else if animationIndex == 2 {
            title = "CurlUp"
            animationView.app.addAnimation(curve: .curveEaseInOut, transition: .transitionCurlUp, duration: 1.0)
        } else if animationIndex == 3 {
            title = "transform.rotation.y"
            animationView.app.addAnimation(keyPath: "transform.rotation.y", fromValue: NSNumber(value: 0), toValue: NSNumber(value: CGFloat.pi), duration: 1.0)
        } else if animationIndex == 4 {
            title = "Shake"
            animationView.app.shake(times: 10, delta: 0, duration: 0.1)
        } else if animationIndex == 5 {
            title = "Alpha"
            animationView.app.fade(alpha: 0, duration: 1.0) { [weak self] _ in
                self?.animationView.app.fade(alpha: 1.0, duration: 1.0)
            }
        } else if animationIndex == 6 {
            title = "Rotate"
            animationView.app.rotate(degree: 180, duration: 1.0)
        } else if animationIndex == 7 {
            title = "Scale"
            animationView.app.scale(scaleX: 0.5, scaleY: 0.5, duration: 1.0) { [weak self] _ in
                self?.animationView.app.scale(scaleX: 2.0, scaleY: 2.0, duration: 1.0)
            }
        } else if animationIndex == 8 {
            title = "Move"
            let point = animationView.frame.origin
            animationView.app.move(point: CGPoint(x: 10, y: 10), duration: 1.0) { [weak self] _ in
                self?.animationView.app.move(point: point, duration: 1.0)
            }
        } else if animationIndex == 9 {
            title = "Frame"
            let frame = animationView.frame
            animationView.app.move(frame: CGRect(x: 10, y: 10, width: 50, height: 50), duration: 1.0) { [weak self] _ in
                self?.animationView.app.move(frame: frame, duration: 1.0)
            }
        } else if animationIndex == 10 {
            title = "切换动画"
            animationIndex = 0
        }
        
        sender.setTitle(title, for: .normal)
    }
    
    @objc func onDrag(_ sender: UIButton) {
        if !animationView.app.dragEnabled {
            animationView.app.dragEnabled = true
            animationView.app.dragLimit = CGRect(x: 0, y: 0, width: APP.screenWidth, height: APP.screenHeight - APP.topBarHeight)
        } else {
            animationView.app.dragEnabled = false
        }
    }
    
}

class TestAnimationView: UIView {
    
    var transitionType: Int = 0
    var edge: UIRectEdge = .bottom
    
    lazy var bottomView: UIView = {
        let result = UIView()
        result.backgroundColor = .white
        return result
    }()
    
    init(transitionType: Int) {
        super.init(frame: .zero)
        self.transitionType = transitionType
        if transitionType == 6 || transitionType == 9 {
            let edges: [UIRectEdge] = [.top, .left, .right, .bottom]
            edge = edges.randomElement()!
        }
        if transitionType > 8 {
            backgroundColor = .clear
        } else {
            backgroundColor = .app.color(hex: 0x000000, alpha: 0.5)
        }
        
        addSubview(bottomView)
        if transitionType == 6 || transitionType == 9 {
            switch edge {
            case .top:
                bottomView.app.layoutChain.horizontal().top().height(APP.screenHeight / 2)
            case .left:
                bottomView.app.layoutChain.vertical().left().width(APP.screenWidth / 2)
            case .right:
                bottomView.app.layoutChain.vertical().right().width(APP.screenWidth / 2)
            default:
                bottomView.app.layoutChain.horizontal().bottom().height(APP.screenHeight / 2)
            }
        } else {
            bottomView.app.layoutChain.center().width(300).height(200)
        }
        
        app.addTapGesture { [weak self] _ in
            guard let self = self else { return }
            if self.transitionType > 8 {
                self.app.viewController?.dismiss(animated: true)
                return
            }
            
            if self.transitionType == 6 {
                self.app.setPresentTransition(.dismiss, contentView: self.bottomView, edge: edge, completion: nil)
            } else if self.transitionType == 7 {
                self.app.setAlertTransition(.dismiss, completion: nil)
            } else {
                self.app.setFadeTransition(.dismiss, completion: nil)
            }
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func show(in viewController: UIViewController) {
        if transitionType > 8 {
            let wrappedVC = app.wrappedTransitionController(true)
            if transitionType == 9 {
                wrappedVC.app.setPresentTransition(edge: edge)
            } else if transitionType == 10 {
                wrappedVC.app.setAlertTransition(nil)
            } else {
                wrappedVC.app.setFadeTransition(nil)
            }
            viewController.present(wrappedVC, animated: true)
            return
        }
        
        app.transition(to: viewController, pinEdges: true)
        if transitionType == 6 {
            app.setPresentTransition(.present, contentView: self.bottomView, edge: edge, completion: nil)
        } else if transitionType == 7 {
            app.setAlertTransition(.present, completion: nil)
        } else {
            app.setFadeTransition(.present, completion: nil)
        }
    }
    
}

class TestAnimationChildController: UIViewController, ViewControllerProtocol {
    
    var transitionType: Int = 0
    
    lazy var bottomView: UIView = {
        let result = UIView()
        result.backgroundColor = .white
        return result
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        app.navigationBarHidden = true
        view.backgroundColor = .clear
        view.app.addTapGesture { [weak self] _ in
            self?.app.close()
        }
        
        view.addSubview(self.bottomView)
        if transitionType == 0 || transitionType == 3 {
            bottomView.app.layoutChain.horizontal().bottom().height(APP.screenHeight / 2)
        } else {
            bottomView.app.layoutChain.center().width(300).height(200)
        }
        
        let button = UIButton()
        button.setTitleColor(.black, for: .normal)
        button.setTitle(self.navigationController != nil ? "支持push" : "不支持push", for: .normal)
        bottomView.addSubview(button)
        button.app.addTouch { [weak self] _ in
            let vc = TestAnimationChildController()
            vc.transitionType = self?.transitionType ?? 0
            self?.navigationController?.pushViewController(vc, animated: true)
        }
        button.app.layoutChain.center()
    }
    
    func show(in viewController: UIViewController) {
        if transitionType == 0 {
            app.setPresentTransition(nil)
        } else if transitionType == 1 {
            app.setAlertTransition(nil)
        } else {
            app.setFadeTransition(nil)
        }
        viewController.present(self, animated: true)
    }
    
    func showNav(in viewController: UIViewController) {
        let nav = UINavigationController(rootViewController: self)
        if transitionType == 3 {
            nav.app.setPresentTransition(nil)
        } else if transitionType == 4 {
            nav.app.setAlertTransition(nil)
        } else {
            nav.app.setFadeTransition(nil)
        }
        viewController.present(nav, animated: true)
    }
    
}
